package edu.jhuapl.sd.sig.mmtc.cfg;

import edu.jhuapl.sd.sig.mmtc.RequiresSpice;
import edu.jhuapl.sd.sig.mmtc.app.MmtcException;
import edu.jhuapl.sd.sig.mmtc.util.TimeConvert;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;
import org.mockito.Mockito;

import edu.jhuapl.sd.sig.mmtc.util.Environment;
import spice.basic.KernelDatabase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Provides unit tests of TimeCorrelationXmlPropertiesConfig's ability to reject kernal path configurations that
 * contain consecutive or trailing commas.
 *
 * TimeCorrelationXmlPropertiesConfig doesn't make this particularly easy to test. Without larger refactoring, the best
 * we can to is mock the environment variable that specifies the path where TimeCorrelationXmlPropertiesConfig searches
 * for the config file so we can provide test versions of the default config file.
 *
 * However, note that this approach is a little risky: if we accidentally code an invalid test path, MMTC will still
 * find the default (non-test version) config file from src/main/resources since that's also on the classpath. The
 * tests will then be using unintented test inputs and giving you false results.
 */
@RequiresSpice
class TimeCorrelationCliAppConfigTests {

    @Test
    @DisplayName("TimeCorrelationAppConfig.getKernelsToLoad baseline/negative test - no exceptions expected")
    void testCorrectCommasInKernelConfigs() throws Exception {
		TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));
		assertDoesNotThrow(
			() -> {
				config.getKernelsToLoad();
			}
		);
	}

    @Test
    @DisplayName("TimeCorrelationAppConfig.getKernelsToLoad trailing commas test")
    void testTrailingCommasInKernelConfigs() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
				.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
				.thenReturn("src/test/resources/ConfigTests/trailingCommas");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));
			MmtcException e = assertThrows(
				MmtcException.class,
				() -> {
					config.getKernelsToLoad();
				}
			);
			assertEquals("Consecutive or trailing commas in kernel configurations", e.getMessage());
		}
	}

    @Test
    @DisplayName("TimeCorrelationAppConfig.getKernelsToLoad consecutive commas test")
    void testConsecutiveCommasInKernelConfigs() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
				.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
				.thenReturn("src/test/resources/ConfigTests/consecutiveCommas");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));
			MmtcException e = assertThrows(
				MmtcException.class,
				() -> {
					config.getKernelsToLoad();
				}
			);
			assertEquals("Consecutive or trailing commas in kernel configurations", e.getMessage());
		}
	}

	@Test
	void testParseVcidGroups() throws MmtcException {
		assertEquals(Arrays.asList(set(1)), MmtcConfig.parseVcidGroups("testkey", "1"));
		assertEquals(Arrays.asList(set(1), set(2)), MmtcConfig.parseVcidGroups("testkey", "1; 2"));

		assertEquals(Arrays.asList(set(1,2)), MmtcConfig.parseVcidGroups("testkey", "1,2"));
		assertEquals(Arrays.asList(set(1,2)), MmtcConfig.parseVcidGroups("testkey", "1,2;"));
		assertEquals(Arrays.asList(set(1,2), set(3,4)), MmtcConfig.parseVcidGroups("testkey", "1,2; 3,4"));
		assertEquals(Arrays.asList(set(1,2), set(3,4), set(3,5)), MmtcConfig.parseVcidGroups("testkey", "1,2; 3,4; 3,5"));

		assertEquals(Arrays.asList(set(0,6,7), set(5)), MmtcConfig.parseVcidGroups("testkey", "0,6,7; 5"));

		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", ""); });
		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", " "); });
		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", ";"); });
		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", " ; "); });
		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", " ; 2"); });

		assertThrows(Exception.class, () -> { MmtcConfig.parseVcidGroups("testkey", "1,2 ; a,b"); });
	}

	@Test
	void testGetTkSclkFineTickModulusWithOverride() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
					.thenReturn("src/test/resources/ConfigTests/unusualSclkModulusOverride");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));

			assertEquals(12345, config.getTkSclkFineTickModulus());
		}
	}

	@Test
	void testGetTkSclkFineTickModulusNoOverride() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
					.thenReturn("src/test/resources/ConfigTests/noSclkModulusOverride");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));

			TimeConvert.loadSpiceLib();
			KernelDatabase.load("src/test/resources/nh_kernels/lsk/naif0012.tls");
			KernelDatabase.load("src/test/resources/nh_kernels/sclk/new-horizons_1454.tsc");

			assertEquals(50000, config.getTkSclkFineTickModulus());
		} finally {
			TimeConvert.unloadSpiceKernels();
		}
	}

	@Test
	void testGetDefaultValuesForTkOscTempAndParmWindowSec() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
					.thenReturn("src/test/resources/ConfigTests/noTkOscTempOrParmWindow");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));

			assertEquals(600, config.getTkOscTempWindowSec());
			assertEquals(600, config.getTkParmWindowSec());
		}
	}

    @Test
    void testCatchesMissingRequiredConfigKeys() throws Exception {
        try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
            mockedEnvironment
                    .when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
                    .thenReturn("src/test/resources/ConfigTests/missingKeys");

            ArrayList<String> expectedMissingVals = new ArrayList<>();
            expectedMissingVals.add("spacecraft.id");
            expectedMissingVals.add("spice.kernel.sclk.baseName");
            expectedMissingVals.add("spice.kernel.sclk.separator");
            expectedMissingVals.add("spice.kernel.sclk.uniqueKernelCounters");

            MmtcException resultingException = assertThrows(
                    MmtcException.class,
                    () -> new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"))
            );

            assertTrue(resultingException.getMessage().startsWith("Failed to validate TimeCorrelationConfigProperties.xml, missing 4 required key(s):"));
            for (String expectedMissingVal : expectedMissingVals) {
                assertTrue(resultingException.getMessage().contains(expectedMissingVal));
            }
        }
    }

	@Test
	void testPassesConfigWithRequiredKeys() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
                    .thenReturn("src/test/resources/ConfigTests/minimalKeys");
			TimeCorrelationRunConfig config = new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"));

			config.validate();
		}
	}

	@Test
	void testFailsConfigDueToMissingSclkScetKey() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
					.thenReturn("src/test/resources/ConfigTests/missingSclkScetKey");

			MmtcException resultingException = assertThrows(
					MmtcException.class,
                    () -> new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"))
			);

			assertEquals("SCLK-SCET operations require the following keys to be set: [product.sclkScetFile.producerId, product.sclkScetFile.scetUtcPrecision]", resultingException.getMessage());
		}
	}

	@Test
	void testFailsConfigDueToMissingUplinkCmdFileKey() throws Exception {
		try (MockedStatic<Environment> mockedEnvironment = Mockito.mockStatic(Environment.class, Mockito.CALLS_REAL_METHODS)) {
			mockedEnvironment
					.when(() -> Environment.getEnvironmentVariable("TK_CONFIG_PATH"))
					.thenReturn("src/test/resources/ConfigTests/missingUplinkCmdFileKey");

			MmtcException resultingException = assertThrows(
					MmtcException.class,
                    () -> new TimeCorrelationRunConfig(new TimeCorrelationCliInputConfig("2020-001T00:00:00", "2020-001T23:59:59"))
			);

			assertEquals("Uplink command file operations require the following keys to be set: [product.uplinkCmdFile.outputDir]", resultingException.getMessage());
		}
	}

	private static Set<Integer> set(Integer... ints) {
		return new HashSet<>(Arrays.asList(ints));
	}
}
